### PR TITLE
Fix the problem that ignoreUriCase does not work when the path type is prefix

### LINKED ISSUE #258: The path prefix match uses a regular pattern, which results in ignoring case not work.
Now: 

prefix match of "/abc" -> regex match of "/abc(\\/.*)?"

Expect: 

prefix match of "/abc" -> prefix match of "/abc/"   and  exact match of "/abc"

### PR DESCRIPTION
**Pull request by Charlie17Li**
### Ⅰ. Describe what this PR did

fix prefix path match bug

### Ⅱ. Does this pull request fix one issue?

fixes #258 

### Ⅴ. Special notes for reviews

1. the `/abc` can match `/abc` and `/abcd` in virtual servcie
2. the `/abc/` can match `/abc/*`, but cannot match `/abc` in virtual servcie
3. but `/abc/` should also match `/abc` in [ingress documetation](https://kubernetes.io/docs/concepts/services-networking/ingress/#examples)
4. so I both convert them(`/abc, /abc/`) to prefix `/abc/` and exact `/abc`

### CODE REVIEW

**Comment by johnlanni** [comment]
cc @SpecialYang

**Review by SpecialYang — CHANGES_REQUESTED** [review]
We respect the spec of Ingress not VirtualService.

File: pkg/ingress/kube/ingressv1/controller.go [hunk 1]
```diff
@@ -539,9 +542,22 @@ func (c *controller) ConvertHTTPRoute(convertOptions *common.ConvertOptions, wra
 							MatchType: &networking.StringMatch_Prefix{Prefix: path},
 						}
 					} else {
-						path = strings.TrimSuffix(path, "/")
+						// the path maybe prefix /xxx or prefix /xxx/
+						// we both convert it to prefix /xxx/ and exact /xxx
+						var extraExactPath string
+						if path[len(path)-1] != '/' { // prefix /xxx
```
Comment (johnlanni) [inline pkg/ingress/kube/ingressv1/controller.go:548]: Why not trim the possible suffix "/" first, so the logic will look clearer

**Review by johnlanni — CHANGES_REQUESTED** [review]

File: pkg/ingress/kube/ingressv1/controller.go [hunk 2]
```diff
@@ -594,6 +610,14 @@ func (c *controller) ConvertHTTPRoute(convertOptions *common.ConvertOptions, wra
...
 		}
 
+		for _, httpPath := range rule.HTTP.Paths {
+			convertPath(httpPath)
+		}
+
+		for _, httpPath := range extraExactPaths {
+			convertPath(httpPath)
+		}
+
```
Comment (johnlanni) [inline pkg/ingress/kube/ingressv1/controller.go:620]: It's a bit tricky here, you'd better convert the ingress path in one step

**Review by johnlanni — CHANGES_REQUESTED** [review]

**Comment by Charlie17Li** [comment]
> We respect the spec of Ingress not VirtualService.

Yes, so I convert prefix `/abc/` to prefix `/abc/` + exact `/abc`, otherwise it will not follow ingress spec

File: pkg/ingress/kube/ingressv1/controller.go (hunk 1 above)
Comment (Charlie17Li) [reply pkg/ingress/kube/ingressv1/controller.go]: done

**Comment by johnlanni** [comment]
> It has already been supported in istio-envoy [istio/istio#38233](https://github.com/istio/istio/pull/38233). Not sure whether the current verison of higress supports.

Not supported now, higress is based on envoy 1.20. Wait for the overall upgrade to support this

File: pkg/ingress/kube/annotations/canary.go [hunk 3]
```diff
@@ -133,39 +133,41 @@ func ApplyByHeader(canary, route *networking.HTTPRoute, canaryIngress *Ingress)
...
-			canaryConfig.Header: {
-				MatchType: &networking.StringMatch_Exact{
-					Exact: "always",
-				},
-			},
-		}
-		if canaryConfig.HeaderValue != "" {
-			canary.Match[0].Headers = map[string]*networking.StringMatch{
+	for _, match := range canary.Match {
+		if canaryConfig.Header != "" {
```
Comment (johnlanni) [inline pkg/ingress/kube/annotations/canary.go:137]: This `if` should be above.

**Review by johnlanni — CHANGES_REQUESTED** [review]

File: pkg/ingress/kube/ingressv1/controller.go [hunk 4]
```diff
@@ -507,47 +506,44 @@ func (c *controller) ConvertHTTPRoute(convertOptions *common.ConvertOptions, wra
 		}
 
 		wrapperHttpRoutes := make([]*common.WrapperHTTPRoute, 0, len(rule.HTTP.Paths))
+
 		for _, httpPath := range rule.HTTP.Paths {
+			var pathType common.PathType
```
Comment (SpecialYang) [inline pkg/ingress/kube/ingressv1/controller.go:511]: Also should add this support for ingress v1beta1.

File: pkg/ingress/kube/ingressv1/controller.go [hunk 5]
```diff
@@ -507,47 +506,44 @@ func (c *controller) ConvertHTTPRoute(convertOptions *common.ConvertOptions, wra
...
-				httpMatch.Uri = &networking.StringMatch{
-					MatchType: &networking.StringMatch_Regex{Regex: httpPath.Path + ".*"},
-				}
-			} else {
-				switch *httpPath.PathType {
-				case ingress.PathTypeExact:
+			// if the pathType is Prefix, it will have 2 httpMatches, the first is the type of Exact and the next is Prefix
+			// so if the pathType is Prefix, the wrapperHttpRoute.OriginPathType is still Prefix, though it has been assigned twice
+			httpMatches := c.generateHttpMatches(pathType, httpPath.Path, wrapperVS)
+			for _, httpMatch := range httpMatches {
```
Comment (SpecialYang) [inline pkg/ingress/kube/ingressv1/controller.go:533]: Here no need to traverse the httpMatches. We treat these two (at most) routes as one route for RouteCache to show for console.

The users of console still get the only one route, just the prefix match.

Comment (Charlie17Li) [reply pkg/ingress/kube/ingressv1/controller.go:533]: I'm not sure if I misunderstood your point🤔, but my understanding is to not traverse through httpMatches, but to only keep the prefix match. 

But for example, if it's `prefix=/abc` or `prefix=/abc/` in Ingress, and we only keep the prefix match, then in the `wrapperHttpRoute.HTTPRoute.Match`, there will only be `prefix=/abc/`. In the current Higress version, this maybe not match the request `path=/abc`.

File: pkg/ingress/kube/ingressv1/controller.go (hunk 4 above)
Comment (Charlie17Li) [reply pkg/ingress/kube/ingressv1/controller.go:511]: ok

File: pkg/ingress/kube/ingress/controller.go [hunk 6]
```diff
@@ -526,46 +525,43 @@ func (c *controller) ConvertHTTPRoute(convertOptions *common.ConvertOptions, wra
 
 		wrapperHttpRoutes := make([]*common.WrapperHTTPRoute, 0, len(rule.HTTP.Paths))
 		for _, httpPath := range rule.HTTP.Paths {
+			var pathType common.PathType
+			if wrapper.AnnotationsConfig.NeedRegexMatch() {
```
Comment (SpecialYang) [inline pkg/ingress/kube/ingress/controller.go:529]: Here we can determine the value of original path.

Comment (Charlie17Li) [reply pkg/ingress/kube/ingress/controller.go:529]: get, so I should convert the original path to httpMatch here and remove generateHttpMatches func. then we don't need to tarverse http matchs below.

File: pkg/ingress/kube/ingress/controller.go [hunk 7]
```diff
@@ -532,40 +531,25 @@ func (c *controller) ConvertHTTPRoute(convertOptions *common.ConvertOptions, wra
...
-							MatchType: &networking.StringMatch_Regex{Regex: regexp.QuoteMeta(path) + common.PrefixMatchRegex},
-						}
-					}
+					pathType = common.Prefix
 				}
 			}
-			wrapperHttpRoute.OriginPath = path
-			wrapperHttpRoute.HTTPRoute.Match = []*networking.HTTPMatchRequest{httpMatch}
+			wrapperHttpRoute.OriginPathType = pathType
+			wrapperHttpRoute.OriginPath = httpPath.Path
```
Comment (SpecialYang) [inline pkg/ingress/kube/ingress/controller.go:547]: For backward compatibility, we should assgin the trimmed path to OriginPath. i.e, the path `/path/` should be `/path`.

```
case ingress.PathTypePrefix:
	if path != "/" {
      path = strings.TrimSuffix(path, "/")
}
```

Comment (Charlie17Li) [reply pkg/ingress/kube/ingress/controller.go:547]: get

File: pkg/ingress/kube/ingress/controller.go [hunk 8]
```diff
@@ -532,40 +531,25 @@ func (c *controller) ConvertHTTPRoute(convertOptions *common.ConvertOptions, wra
...
+
+			httpMatches := c.generateHttpMatches(pathType, httpPath.Path, wrapperVS)
+			for _, httpMatch := range httpMatches {
+				wrapperHttpRoute.HTTPRoute.Match = append(wrapperHttpRoute.HTTPRoute.Match, httpMatch)
```
Comment (SpecialYang) [inline pkg/ingress/kube/ingress/controller.go:551]: Can we simplely assgin httpMatches to `wrapperHttpRoute.HTTPRoute.Matc` ?

Comment (Charlie17Li) [reply pkg/ingress/kube/ingress/controller.go]: get!

File: pkg/ingress/kube/ingressv1/controller.go [hunk 9]
```diff
@@ -507,47 +506,37 @@ func (c *controller) ConvertHTTPRoute(convertOptions *common.ConvertOptions, wra
...
 				HTTPRoute:     &networking.HTTPRoute{},
 				WrapperConfig: wrapper,
 				Host:          rule.Host,
 				ClusterId:     c.options.ClusterId,
 			}
-			httpMatch := &networking.HTTPMatchRequest{}
 
-			path := httpPath.Path
+			var pathType common.PathType
+			var originPath string
```
Comment (SpecialYang) [inline pkg/ingress/kube/ingressv1/controller.go:519]: originPath := httpPath.Path

File: pkg/ingress/kube/ingressv1/controller.go [hunk 10]
```diff
@@ -507,47 +506,37 @@ func (c *controller) ConvertHTTPRoute(convertOptions *common.ConvertOptions, wra
...
 			if wrapper.AnnotationsConfig.NeedRegexMatch() {
-				wrapperHttpRoute.OriginPathType = common.Regex
-				httpMatch.Uri = &networking.StringMatch{
-					MatchType: &networking.StringMatch_Regex{Regex: httpPath.Path + ".*"},
-				}
+				pathType = common.Regex
+				originPath = httpPath.Path
```
Comment (SpecialYang) [inline pkg/ingress/kube/ingressv1/controller.go:522]: Can remove.

File: pkg/ingress/kube/ingressv1/controller.go [hunk 11]
```diff
@@ -507,47 +506,37 @@ func (c *controller) ConvertHTTPRoute(convertOptions *common.ConvertOptions, wra
...
-					wrapperHttpRoute.OriginPathType = common.Prefix
-					// borrow from implement of official istio code.
-					if path == "/" {
-						wrapperVS.ConfiguredDefaultBackend = true
-						// Optimize common case of / to not needed regex
-						httpMatch.Uri = &networking.StringMatch{
-							MatchType: &networking.StringMatch_Prefix{Prefix: path},
-						}
+					pathType = common.Prefix
+					if httpPath.Path != "/" {
```
Comment (SpecialYang) [inline pkg/ingress/kube/ingressv1/controller.go:530]: If path is '/', the origin path will be empty. So you should assgin first, just like mentioned above.

File: pkg/ingress/kube/ingressv1/controller.go (hunk 9 above)
Comment (Charlie17Li) [reply pkg/ingress/kube/ingressv1/controller.go]: get

**Review by SpecialYang — APPROVED** [review]
Looks good. Thanks.

**Review by johnlanni — APPROVED** [review]

**Pull request merged**